clock.go

```diff
@@ -21,33 +21,57 @@ type Clock interface {
 	// NewTimer creates a new Timer that will send the current time
 	// on its channel after at least duration d.
 	NewTimer(d time.Duration) Timer
-}
 
-// Alarm returns a channel that will have the time sent on it at some point
-// after the supplied time occurs.
-//
-// This is short for c.After(t.Sub(c.Now())).
-func Alarm(c Clock, t time.Time) <-chan time.Time {
-	return c.After(t.Sub(c.Now()))
+	// At waits for the time to pass and then sends the
+	// current time on the returned channel.
+	At(t time.Time) <-chan time.Time
+
+	// AtFunc waits for the time to pass and then calls f in its own goroutine.
+	// It returns an Alarm that can be used to cancel the call using its Stop method.
+	AtFunc(t time.Time, f func()) Alarm
+
+	// NewAlarm creates a new Alarm that will send the current time
+	// on its channel at or after time t has passed.
+	NewAlarm(t time.Time) Alarm
 }
 
-// The Timer type represents a single event.
-// A Timer must be created with AfterFunc.
+// Timer type represents a single event.
+// Timers must be created with AfterFunc or NewTimer.
 // This interface follows time.Timer's methods but provides easier mocking.
 type Timer interface {
-	// When the Timer expires, the current time will be sent on the
-	// channel returned from Chan, unless the Timer was created by
+	// When the timer expires, the current time will be sent on the
+	// channel returned from Chan, unless the timer was created by
 	// AfterFunc.
 	Chan() <-chan time.Time
 
 	// Reset changes the timer to expire after duration d.
 	// It returns true if the timer had been active, false if
 	// the timer had expired or been stopped.
-	Reset(time.Duration) bool
+	Reset(d time.Duration) bool
 
 	// Stop prevents the Timer from firing. It returns true if
 	// the call stops the timer, false if the timer has already expired or been stopped.
 	// Stop does not close the channel, to prevent a read
 	// from the channel succeeding incorrectly.
 	Stop() bool
 }
+
+// Alarm type represents a single event.
+// Alarms must be created with AtFunc or NewAlarm.
+type Alarm interface {
```

Comment:
**thought:** This abstraction feels a bit  weird: It is too close to Timer abstraction (with only `Reset` signature as a differencies.

`Stop()` isn't a correct abstraction, I think, since we don't "stop" an alarm. Stopping a timer has a sense since when restart it, we still have time to go.

**suggestion:** I think the right abstraction here is something like we have on our phone.

* `Timer` can be stopped and resume
* `Alarm` can be enabled or disabled. In future, they may become periodic (set an alarm everyday, at 5...

`Reset` feels weird to me. Why having a reset timer when we can just throw up a new timer ? 

However, I don't have the big picture, but with those names it is a little confusing (timer and alarm are implemented on our phone and follows well known standard. If our abstraction isn't that close of the alarm stuff on our phone, maybe our naming is confusing :D )

Reply:
Desktop isn't a desk top, although we may borrow things, skeuomorphic abstractions do not need to be identical, only familiar.

In this case, it is important to give them different names, but functionally, they are very similar mechanisms. Same with the clock app on your phone: Alarms (alert at a point in time, with optional repeating), Timers (alert after a duration of elapsed time with a manual repeat), Stopwatch (a clock with a non standard epoch) and the Clock itself. An alarm is the closest abstraction we have.

For the Stop and Reset interface, I would argue, these are the same concept, they share abstractions, Go is the limiting factor here.

Reply:
I agree on the first point (abstraction - skeuomorphic or less exotic one :p ) don't need to be indentical, only familiar. 

This is an interesting question to state if those are familiar. I stated that put in that way, Alarms and Timer feels weird to me (aka unfamiliar).

I understand the need (having something similar in functionality but with a different name). And I understand the choice of Alarm as a name, as explained by you.

However, my point of view, when I try to adopt "mundane" opinion, is that alarm and timer, besides the fact there are signal emitted over time, will evolves with slightly different path. And since its mundane opinion (or I believe it is), future developer may misunderstand the Alarm here (which is an AbsoluteTime Timer, where actual Timer is a RelativeTime timer, if we choose other sets of abstraction).

However that is only thoughts, so no worries, let's see ;) Thank you for your detailed answer. Agree to disagree :)